src/data/providers/balancer-api/modules/nested-pool-state/index.ts

```diff
@@ -0,0 +1,265 @@
+                type: this.mapPoolType(pool.type),
+                level: 1,
+                tokens: pool.tokens.map((t) => {
+                    const minimalToken: MinimalToken = {
+                        address: t.address,
+                        decimals: t.decimals,
+                        index: t.index,
+                    };
+                    return minimalToken;
+                }),
+            },
+        ];
+
+        pool.tokens.forEach((token) => {
+            // Token represents nested pools only nested if they have a pool property
+            if (token.pool === undefined) return;
+
+            /**
+             * TODO: Current development API has included BPT token in nested pools,
+             * but they are still deciding if that will be reverted.
+             * If they include it -> remove commented code
+             * If they don't -> uncomment code to apply hack
+             */
+            // // hack to add BPT token missing from nested pools on API result
+            // let missingBPTIndex = 0;
+            // const sortedIndexes = token.pool.tokens.map((t) => t.index).sort();
+            // for (let i = 0; i < token.pool.tokens.length + 1; i++) {
+            //     if (i === token.pool.tokens.length || sortedIndexes[i] !== i) {
+            //         missingBPTIndex = i;
+            //     }
+            // }
+            // token.pool.tokens.splice(missingBPTIndex, 0, {
+            //     index: missingBPTIndex,
+            //     name: 'BPT',
+            //     symbol: 'BPT',
+            //     address: token.pool.address,
+            //     decimals: 18,
+            // });
+
+            // map API result to NestedPool
+            pools.push({
+                id: token.pool.id,
+                address: token.pool.address,
+                level: 0,
+                type: this.mapPoolType(token.pool.type),
+                tokens: token.pool.tokens.map((t) => {
+                    const minimalToken: MinimalToken = {
+                        address: t.address,
+                        decimals: t.decimals,
+                        index: t.index,
+                    };
+                    return minimalToken;
+                }),
+            });
+        });
+
+        const mainTokens = pool.allTokens
+            .filter((t) => t.isMainToken)
+            .map((t) => {
+                return {
+                    address: t.address,
+                    decimals: t.decimals,
+                };
+            });
+
+        return {
+            pools,
+            mainTokens,
+        } as NestedPoolState;
+    };
+
+    mapPoolType = (type: string): PoolType => {
+        switch (type) {
+            case 'WEIGHTED':
+                return PoolType.Weighted;
+            case 'PHANTOM_STABLE':
+                return PoolType.ComposableStable;
+            default:
+                throw new Error(`Unsupported pool type ${type}`);
+        }
+    };
```

Comment:
I'd like to highlight this method. I added it in order to separate pool type on API vs b-sdk. Alongside this, I changed `type` from `String` to `PoolType` within `NestedPool`. 
This is somewhat different from the pure join approach, because there we split `PoolState` and `PoolStateInput` to handle this difference on "internal" vs "external" objects.
Following pure join approach, I'd expect to have a `NestedPoolInput` that has `type` as `String` and only within the  `query` method I'd convert it to `NestedPool` with `type` as `PoolType`, but that seems like too much overhead for me 😬